### PR TITLE
Adapt algorithm parameter sequence changes from C++17

### PR DESCRIPTION
**Pull request by hkaiser**
- rename `inner_product` to transform_reduce
- align parameter sequence for `transform_reduce` and `transform_*_scan` algorithms

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -796,6 +796,14 @@ if(HPX_WITH_BOOST_CHRONO_COMPATIBILITY)
   hpx_add_config_define(HPX_HAVE_BOOST_CHRONO_COMPATIBILITY)
 endif()
 
+# HPX_WITH_TRANSFORM_REDUCE_COMPATIBILITY: introduced in V1.0.0
+hpx_option(HPX_WITH_TRANSFORM_REDUCE_COMPATIBILITY BOOL
+    "Enable old overloads for transform_reduce and inner_product (default: OFF)"
+    OFF ADVANCED)
```
Comment (sithhell) [inline CMakeLists.txt]: Can we have this on by default?

File: hpx/parallel/algorithms/inner_product.hpp [hunk 2]
```diff
@@ -290,22 +81,9 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
     inner_product(ExPolicy&& policy, InIter1 first1, InIter1 last1,
         InIter2 first2, T init)
     {
```
Comment (sithhell) [inline hpx/parallel/algorithms/inner_product.hpp]: Are the asserts not valid anymore?

File: CMakeLists.txt (hunk 1 above)
Comment (hkaiser) [inline CMakeLists.txt]: Yes, good point.

File: hpx/parallel/algorithms/inner_product.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/parallel/algorithms/inner_product.hpp]: This is a simple forwarding function now. The asserts are checked by the called `transform_reduce` algorithm.

File: docs/manual/build_system/cmake_variables.qbk [hunk 3]
```diff
@@ -95,6 +96,7 @@ The options are split into these categories:
         [[[#build_system.cmake_variables.HPX_WITH_SECURITY] `HPX_WITH_SECURITY:BOOL`][Enable security support via libsodium.]]
         [[[#build_system.cmake_variables.HPX_WITH_STATIC_LINKING] `HPX_WITH_STATIC_LINKING:BOOL`][Compile HPX statically linked libraries (Default: OFF)]]
         [[[#build_system.cmake_variables.HPX_WITH_SYCL] `HPX_WITH_SYCL:BOOL`][Enable sycl support (default: OFF)]]
+        [[[#build_system.cmake_variables.HPX_WITH_TRANSFORM_REDUCE_COMPATIBILITY] `HPX_WITH_TRANSFORM_REDUCE_COMPATIBILITY:BOOL`][Enable old overloads for transform_reduce and inner_product (default: OFF)]]
```
Comment (sithhell) [inline docs/manual/build_system/cmake_variables.qbk]: the documentation should be updated with the latest value as well.

File: hpx/parallel/algorithms/inner_product.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/parallel/algorithms/inner_product.hpp]: Hmmm. Ok, I see the point. No use in duplicating this code. Would have been nice to have errors as early as possible in the call chain.

Comment (hkaiser) [inline hpx/parallel/algorithms/inner_product.hpp]: So, should I duplicate the code - or not?

File: docs/manual/build_system/cmake_variables.qbk (hunk 3 above)
Comment (hkaiser) [inline docs/manual/build_system/cmake_variables.qbk]: Right.

**Comment by hkaiser** [comment]
Can I go ahead and merge this now?

**Comment by hkaiser** [comment]
All review comments have been addressed.

**Pull request merged**